### PR TITLE
Topology: sof-icl-rt711-rt1308-rt715-hdmi: Use demux for platforms with multiple speakers

### LINKED ISSUE #2570: [BUG][CML]" error: ipc error for 0x60010000 size 20" display when aplay via speaker0,2.
**Describe the bug**
Speaker0,2 can't work normally, dmesg show " error: ipc error for 0x60010000 size 20".

**To Reproduce**
1.Boot up system.
2.aplay -Dhw:0,2 -fs16_le -c2 -r 48000 -vv -i wav/48California_Dreaming.wav

**Reproduction Rate**
5/5

**Expected behavior**
Aplay0,2  should work normally.

**Impact**
" error: ipc error for 0x60010000 size 20" display when aplay via speaker0,2.
```
[   43.225747] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x40000 successful
[   43.225770] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x40000 successful
[   43.225776] sof-audio-pci 0000:00:1f.3: period_bytes:0x3fc0
[   43.225778] sof-audio-pci 0000:00:1f.3: periods:4
[   43.225800] sof-audio-pci 0000:00:1f.3: stream_tag 1
[   43.225812] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   43.226331] sof-audio-pci 0000:00:1f.3: error: ipc error for 0x60010000 size 20
[   43.226343] sof-audio-pci 0000:00:1f.3: error: hw params ipc failed for stream 1
[   43.226348] sof-audio-pci 0000:00:1f.3: ASoC: 0000:00:1f.3 hw params failed: -22
[   43.226353]  SDW1-speakers: ASoC: hw_params FE failed -22
[   43.226364] sdw_deprepare_stream: subdevice #0-Playback: inconsistent state state 1
[   43.226379] intel-master sdw-master-1: sdw_deprepare_stream: failed -22
[   43.226396] sof-audio-pci 0000:00:1f.3: pcm: free stream 2 dir 0
[   43.226404] sdw_deprepare_stream: subdevice #0-Playback: inconsistent state state 1
[   43.226415] intel-master sdw-master-1: sdw_deprepare_stream: failed -22
[   43.226886] sof-audio-pci 0000:00:1f.3: pcm: close stream 2 dir 0
[   43.636422] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   43.636449] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   44.136257] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   44.136292] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   46.543261] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x201
```
```
ubuntu@cd-CML-H-rt711-rt1308-rt715:~$ aplay -Dhw:0,2 -r48000 -c2 -fs16_le wav/48K_Let_It_Go.wav -vv -i
Playing WAVE 'wav/48K_Let_It_Go.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
aplay: set_params:1363: Unable to install hw params:
ACCESS:  RW_INTERLEAVED
FORMAT:  S16_LE
SUBFORMAT:  STD
SAMPLE_BITS: 16
FRAME_BITS: 32
CHANNELS: 2
RATE: 48000
PERIOD_TIME: 85000
PERIOD_SIZE: 4080
PERIOD_BYTES: 16320
PERIODS: (4 5)
BUFFER_TIME: 341000
BUFFER_SIZE: 16368
BUFFER_BYTES: 65472
TICK_TIME: 0
```

**Test recipe**

Kernel:https://github.com/thesofproject/linux/commits/topic/sof-dev  commit:45f6424
FW: https://github.com/thesofproject/sof/commits/master commit:  a133d5
Tplg: Same with FW branch,  sof-cml-rt711-rt1308-rt715.tplg
Platform:CML-H Laptop with codec ALC711, 2 x RT1308, RT715 in SDW mode.

[dmesg20.log](https://github.com/thesofproject/sof/files/4347372/dmesg20.log)

### PR DESCRIPTION
**Pull request by bardliao**
We only need demux for two speaker dais case.

Fixes: 5e8de9f "topology: sof-icl-rt711-rt1308-rt715-hdmi: Merge two pipeline with demux"
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

Fixes #2570

### CODE REVIEW

**Comment by RanderWang** [comment]
@bardliao  How about change your title to reflect your change ? "use demux for platforms with more than one speaker"  And commit message may disclose how to do in FW for this case.

**Comment by bardliao** [comment]
> @bardliao so where is the "MONO" coming from?

We will define [PLATFORM](https://github.com/thesofproject/sof/blob/master/tools/topology/CMakeLists.txt#L89) in CMakeLists.txt, and define "MONO" in specific platform m4 file. e.g. [cml-mono.m4](https://github.com/thesofproject/sof/blob/master/tools/topology/platform/intel/cml-mono.m4)

**Comment by bardliao** [comment]
> @bardliao How about change your title to reflect your change ? "use demux for platforms with multiple speakers" And commit message may disclose how to do in FW for this case.

Done

**Comment by plbossart** [comment]
To answer to @juimonen there is an abuse of language. MONO here means single amplifier. That RT1308 is perfectly capable of playing stereo sounds. Not MONO means 2 amps and sometimes 2 speakers and sometimes 4, it's a topology indication.
I've been wondering if it's too late to remove this misleading MONO wording. We'd need to change the topology names used in the kernel soc-acpi matching tables.

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 [hunk 1]
```diff
@@ -53,7 +53,8 @@ PIPELINE_PCM_ADD(sof/pipe-volume-capture.m4,
 
 # Low Latency playback pipeline 3 on PCM 2 using max 2 channels of s32le.
 # Schedule 48 frames per 1000us deadline on core 0 with priority 0
-PIPELINE_PCM_ADD(sof/pipe-volume-demux-playback.m4,
+PIPELINE_PCM_ADD(ifdef(`MONO',`sof/pipe-volume-playback.m4',
+`sof/pipe-volume-demux-playback.m4',)
```
Comment (plbossart) [inline tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: there's already an ifdef on line 61, it'd probably be a good thing to have clearer separation between the cases.
Also we may want to conditionally disable aggregation, as we do e.g. for the kernel.

Comment (plbossart) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @bardliao what I meant was

````
if MONO
   pipelines, etc
else
  if aggregation
      pipelines, etc
  else 
      pipelines, pcm, etc.
  endif
endif
````

Comment (bardliao) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @plbossart Done. But a little bit different to your example.

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 [hunk 2]
```diff
@@ -134,10 +136,15 @@ DAI_ADD(sof/pipe-dai-playback.m4,
...
+ifdef(`NO_AGGREGATION',
 `# playback DAI is ALH(SDW2 PIN2) using 2 periods
 # Buffers use s24le format, with 48 frame per 1000us on core 0 with priority 0
-DAI_ADD_SCHED(sof/pipe-dai-sched-playback.m4,
+DAI_ADD(sof/pipe-dai-playback.m4,
+	4, ALH, 0x202, SDW2-Playback,
+	PIPELINE_SOURCE_4, 2, s24le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)',
+`ifdef(`MONO', `',
+`DAI_ADD_SCHED(sof/pipe-dai-sched-playback.m4,
```
[the remark this answers is not in the document]
Comment (bardliao) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: > the mix of ifdef is really hard to follow. And on line 148 it's rather suspicious to have SDW1-playback with index 0x202. That looks like a copy paste mistake to me, no?

No, it is intentional. ALH0x202 is the 2nd DAI for SDW1-Playback stream.
From line 54 to 79:
```
if NOMO (1 rt1308)
    Create a normal pipeline 3
else
    if NO_AGGREGATION (2 rt1308s but no aggregation)
        Create a normal pipeline 3 and normal pipeline 4
    else (2 rt1308s and aggregation)
        Create a demux pipeline 3 and DAI endpoint 4
```
From line 139 to 162:
```
if NO_AGGREGATION (2 rt1308s, but no aggregation)
    ALH0x202 DAI is for SDW2-Playback stream
else
    if MONO (No ALH0x202 is there is only 1 rt1308)
    else (2 rt1308s and aggregation)
        ALH0x202 is the 2nd DAI for SDW1-Playback stream and connect pipeline 4 source to the demux
```

[the remark this answers is not in the document]
Comment (plbossart) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: ok, but then let's stop using SDW1-playback, it's confusing. In some cases it does mean playback on Link1 and other it's just the playback stream1...

We can fix this in a follow-up, along with the rename of MONO which is misleading - should be SINGLE_AMP

**Comment by plbossart** [comment]
merging to keep firmware and kernel aligned
